### DISCUSSION TITLE
GITHUB_TOKEN is giving me an error

### DISCUSSION THREAD

**Discussion by youssefbenfarhat** [community/community#56752] [Actions] [answered]
### Select Topic Area

Question

### Body

Hello community,

Do you know how to fix this error? It's a problem related to Github Token. I couldn't know how to fix it ?

API_URL="https://api.github.com/repos/${{ github.repository }}/pulls/$PR_NUMBER/reviews?per_page=100"
echo "API URL: $API_URL"
RESPONSE=$(curl -s -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" "$API_URL")

echo "API Response:"
echo "$RESPONSE"

I'm getting this: 

API Response:
{
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest"
}

**Comment by airtower-luna** [comment]
What permissions does the `GITHUB_TOKEN` have in your workflow?

Also, what kind of event? Is it `pull_request` or something else (that can influence permissions, too, though it shouldn't prevent read access).

**Reply by youssefbenfarhat** [reply]
on:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
  pull_request_review:
    types:
      - submitted
  
  pull_request_review_comment:
    types:
      - created
The workflow is triggered for these events. For the GITHUB_TOKEN, how can I know what permissions it has? from what I can see for the actions under the Settings: I can see that it has Allow all actions and for Workflow permissions: read and write permissions.

**Reply by airtower-luna** [reply]
> For the GITHUB_TOKEN, how can I know what permissions it has?

Check the "Set up job" step in the job log, there should be a "GITHUB_TOKEN Permissions" section.

With those triggers your token might be limited to read-only (if the PR is from a fork).

**Reply by youssefbenfarhat** [reply]
This is what I found in the logs: 

2023-05-30T09:02:10.3200418Z ##[group]GITHUB_TOKEN Permissions
2023-05-30T09:02:10.3200939Z Actions: write
2023-05-30T09:02:10.3201224Z Checks: write
2023-05-30T09:02:10.3201544Z Contents: write
2023-05-30T09:02:10.3201801Z Deployments: write
2023-05-30T09:02:10.3202093Z Discussions: write
2023-05-30T09:02:10.3202381Z Issues: write
2023-05-30T09:02:10.3202614Z Metadata: read
2023-05-30T09:02:10.3202917Z Packages: write
2023-05-30T09:02:10.3203192Z Pages: write
2023-05-30T09:02:10.3203432Z PullRequests: write
2023-05-30T09:02:10.3203759Z RepositoryProjects: write
2023-05-30T09:02:10.3204100Z SecurityEvents: write
2023-05-30T09:02:10.3204353Z Statuses: write
2023-05-30T09:02:10.3204630Z ##[endgroup]

**Reply by airtower-luna** [reply]
Hm, then the token permissions themselves aren't the problem. Can you provide the *full* context (job) of how that script gets called?

Also, is there a reusable workflow in play?

**Reply by youssefbenfarhat** [reply]
name: Pull Request Approval

on:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
  pull_request_review:
    types:
      - submitted
  
  pull_request_review_comment:
    types:
      - created



jobs:
  approval:
    runs-on: ubuntu-latest

    steps:
      - name: List Pre-installed Tools
        run: ls /usr/bin
      - name: Checkout repository
        uses: actions/checkout@v2


      - name: Run approval check
        run: |
          # Get the pull request number
          echo "$GITHUB_REF"
          PR_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
          API_URL="https://api.github.com/repos/${{ github.repository }}/pulls/$PR_NUMBER/reviews?per_page=100"

          echo "API URL: $API_URL"
          RESPONSE=$(curl -s -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" "$API_URL")

          echo "API Response:"
          echo "$RESPONSE"

**Reply by airtower-luna** [reply]
There's no obvious error. Plenty of things that aren't best practice and you should fix (in particular, avoid using Actions syntax `${{ ... }}` in the `run` script), but I'm not sure if any of them would cause that error. Maybe if the token contains some odd character?

You could try cleaning up a bit, using environment variables instead, and requesting verbose output instead of silence from curl:

```yaml
  - name: Run approval check
    env:
      PR_NUMBER: ${{ github.event.pull_request.number }}
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    run: |
      echo "$GITHUB_REF"
      API_URL="https://api.github.com/repos/${GITHUB_REPOSITORY}/pulls/${PR_NUMBER}/reviews?per_page=100"
      echo "API URL: $API_URL"
      curl -v -H "Authorization: Bearer ${GH_TOKEN}" "${API_URL}"
```

I'm not sure if this will fix anything, but at least it rules out some potential issues. :sweat_smile:

**Reply by youssefbenfarhat** [reply]
How can I check the GITHUB_TOKEN? because when I want to print it, it's displaying ***

**Reply by airtower-luna** [reply]
Theoretically you could encode it in some way that hides it from masking, but you probably shouldn't. :sweat_smile: 

My point above was to use it in a way that rules out the "odd character" problem by using environment variables instead of Actions syntax, so quoting in Bash actually works.

**Reply by youssefbenfarhat** [reply]
Still not working and I don't know what's wrong.

**Reply by airtower-luna** [reply]
I don't know either, can you link the actual job log?

**Reply by youssefbenfarhat** [reply]
I'm using Github Enterprise. Do you think this could be the reason? Do you think it's the same API_URL?

**Reply by airtower-luna** [reply]
That's very, very likely the problem. You'll need to use the API URL for your instance then, and it's no surprise its token doesn't work for `github.com`. :sweat_smile:

**Reply by ChristopherHX** [reply, marked as answer]
Yes you should always use `GITHUB_API_URL` (docs: https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) instead of hardcoding the api url.

for example
```
  - name: Run approval check
    env:
      PR_NUMBER: ${{ github.event.pull_request.number }}
      GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    run: |
      echo "$GITHUB_REF"
      API_URL="${GITHUB_API_URL}/repos/${GITHUB_REPOSITORY}/pulls/${PR_NUMBER}/reviews?per_page=100"
      echo "API URL: $API_URL"
      curl -v -H "Authorization: Bearer ${GH_TOKEN}" "${API_URL}"
```

**Reply by youssefbenfarhat** [reply]
The solution was this:  
API_URL="https://MY-ENTREPRISE/api/v3".
RESPONSE=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
          -H "Accept: application/vnd.github.v3+json" \
          "$API_URL/repos/$REPO_NAME/pulls/$PR_NUMBER/reviews")

**Comment by youssefbenfarhat** [comment]
@airtower-luna @ChristopherHX  do you know how can I check the number of the approvers? I have a team composed of 3(dev1, dev2,dev3). I want to check if they approved the pull request or not and get the number of approvals. 
I wrote this script nut it's always returning 0. 


RESPONSE=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
          -H "Accept: application/vnd.github.v3+json" \
          "$API_URL/repos/$REPO_NAME/pulls/$PR_NUMBER/reviews")
        
echo "$RESPONSE"
TEAM_APPROVALS=$(echo "$RESPONSE" | grep -E '^\s+\"state\": \"APPROVED\"' | grep -E '^\s+\"user\": \{.*\"login\": \"(Dev1,Dev2,De3)\"' | wc -l)



Also when I print this script: 
echo "$RESPONSE" | grep -E '^\s+\"state\": \"APPROVED\"'

IT'S RETURNING THIS: 
 "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",
    "state": "APPROVED",

**Reply by airtower-luna** [reply]
`grep` is pretty certainly not the right tool there. I'd write a tiny Python script that parses the JSON, and then iterates over the reviews (check if the user is one you care about, then the state, collect results however you like). If you're a `jq` wizard you might be able to do it with that, too.

**Reply by youssefbenfarhat** [reply]
Do you know how to do it with Python? Here I want to pass the output of the response to the the job as an input and then parse the JSON. But I couldn't know how to do it. Do you have any idea?
 - name: Run approval check
        id: github-url
        env:
          PR_NUMBER: ${{ github.event.pull_request.number }}
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          # Get the pull request number
          echo "$GITHUB_REF"
          API_URL="https://ENTREPRISE/api/v3"
          REPO_NAME="${{ github.repository }}"
          echo "Calling GitHub API for $REPO_NAME"
          
          RESPONSE=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
          -H "Accept: application/vnd.github.v3+json" \
          "$API_URL/repos/$REPO_NAME/pulls/$PR_NUMBER/reviews")
        
          #echo "$RESPONSE"
      
          echo "RESPONSE=echo "$RESPONSE"" >> "$GITHUB_OUTPUT"

        
      - name: Python script
        run: |
          import os
          
          print ("HELLO WORLD",${{join(steps.github-url.outputs.*, '\n')}})
          print(os.environ['PATH'])
        shell: python

**Reply by airtower-luna** [reply]
Trying to inject the JSON into your Python script with Actions syntax (`${{ ... }}`) is definitely going to mess things up, because that's inserted *literally*. The approach I'd choose is:

* Create a temporary file, store the path in an output.
* Redirect the `curl` output to that file, instead of using `$(...)` syntax to assign a variable.
* In the Python step, read that file. Pass the file path to the script as an environment variable.

Approximately this (untested):

```yaml
- name: Run approval check
  id: github-url
  run: |
    # ...
    json_file="$(mktemp)"
    echo "json_file=${json_file}" >> "${GITHUB_OUTPUT}"
    curl [all your options here] > "${json_file}"
- name: Python script
  shell: python
  env:
    JSON_FILE: ${{ steps.github-url.outputs.json_file }}
  run: |
    import json
    import os

    with open(os.environ['JSON_FILE']) as fh:
        resp = json.load(fh)
```

**Reply by youssefbenfarhat** [reply]
At the end I could do it with Python. I did it with with jq. At the begging I was able to get the list of all the the approvers with this GET request RESPONSE=$(curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" \
          -H "Accept: application/vnd.github.v3+json" \
          "$API_URL/repos/$REPO_NAME/pulls/$PR_NUMBER/reviews") 

But for example, today I had 4 new approvals. I can see only one when I do echo "$RESPONSE". I don't know why.

**Reply by youssefbenfarhat** [reply]
I had many approvals and commits before. I'm thinking maybe the problem is the paginator. I don't know if it exists or not.
